### ISSUE TITLE
ray.get memory is "slow"

### ISSUE THREAD

**Issue by yaroslavvb** [ray-project/ray#1614]
TLDR; There's something about memory allocated by ray.get that makes it hard to get that data into PyTorch/TensorFlow fast enough. Numpy arrays can take 9-12ms to ingest, but numpy arrays coming from Ray take 45ms

Deduping from [discussion](https://groups.google.com/forum/#!topic/ray-dev/OWOrmalTxro) in ray-dev to take advantage of post 1990s word-editing technology.

To get a scale of ideal performance, consider that V100 computes Resnet-50 gradient in 100ms for batch 32, and this gradient is about 100MB in size. Waiting for 45 ms to get Ray parameter values into the framework would drop GPU utilization by 30%

https://github.com/diux-dev/cluster/blob/master/yuxin_numpy/tf_numpy_benchmark.py

```
regular numpy array -> PyTorch GPU memory
python tf_numpy_benchmark.py --benchmark=pytorchgpu_from_numpy --allocator=numpy --num-iters=51
pytorch_from_numpy            :   8.1 GB/sec, min: 12.34, median: 12.37, mean: 12.38
pytorch_from_numpy            :   8.1 GB/sec, min: 12.34, median: 12.39, mean: 12.39


regular numpy array -> TensorFlow GPU memory
python tf_numpy_benchmark.py --benchmark=feed_gpu_tensor --allocator=numpy --num-iters=51
feed_gpu_tensor               :   2.1 GB/sec, min: 47.08, median: 47.43, mean: 47.53
feed_gpu_tensor               :   2.1 GB/sec, min: 47.36, median: 47.94, mean: 47.97

regular numpy array -> TensorFlow GPU memory (using tcmalloc)
sudo apt-get install -y google-perftools
export LD_PRELOAD="/usr/lib/libtcmalloc.so.4" 
python tf_numpy_benchmark.py --benchmark=feed_gpu_tensor --allocator=numpy --num-iters=51
feed_gpu_tensor               :   7.9 GB/sec, min: 12.72, median: 12.82, mean: 12.82
feed_gpu_tensor               :   7.9 GB/sec, min: 12.73, median: 13.07, mean: 13.00

page-locked numpy array -> PyTorch GPU memory
python tf_numpy_benchmark.py --benchmark=pytorchgpu_from_numpy --allocator=tfgpu --num-iters=51
pytorch_from_numpy            :  11.1 GB/sec, min:  9.03, median:  9.03, mean:  9.03
pytorch_from_numpy            :  11.1 GB/sec, min:  9.02, median:  9.03, mean:  9.03

page-locked numpy array -> TensorFlow GPU memory
python tf_numpy_benchmark.py --benchmark=feed_gpu_tensor --allocator=tfgpu --num-iters=51
feed_gpu_tensor               :  10.4 GB/sec, min:  9.61, median:  9.70, mean:  9.71
feed_gpu_tensor               :  10.4 GB/sec, min:  9.58, median:  9.74, mean:  9.75


ray numpy array -> PyTorch GPU memory
python tf_numpy_benchmark.py --benchmark=pytorchgpu_from_numpy --allocator=ray --num-iters=51
pytorchgpu_from_numpy         :   2.2 GB/sec, min: 45.56, median: 45.84, mean: 45.91
pytorchgpu_from_numpy         :   2.2 GB/sec, min: 45.73, median: 45.90, mean: 45.97

ray numpy array -> TensorFlow GPU memory
python tf_numpy_benchmark.py --benchmark=feed_gpu_tensor --allocator=ray --num-iters=51
feed_gpu_tensor               :   1.1 GB/sec, min: 91.34, median: 91.84, mean: 91.76
feed_gpu_tensor               :   1.1 GB/sec, min: 89.31, median: 89.80, mean: 90.01

ray numpy array -> TensorFlow GPU memory (using tcmalloc)
sudo apt-get install -y google-perftools
export LD_PRELOAD="/usr/lib/libtcmalloc.so.4" 
python tf_numpy_benchmark.py --benchmark=feed_gpu_tensor --allocator=ray --num-iters=51
feed_gpu_tensor               :   2.0 GB/sec, min: 51.05, median: 51.40, mean: 51.49
feed_gpu_tensor               :   2.1 GB/sec, min: 47.71, median: 48.40, mean: 48.70
```

**Comment by robertnishihara** [comment]
Is the issue fully explained by https://groups.google.com/d/msg/ray-dev/pNsxWI-iSyI/YOwKBqEgAwAJ?

Numpy arrays created by `ray.get` are read-only (since the arrays are backed by shared memory), and when a pytorch tensor is initialized from a read-only array, the array is copied. Does that explain the full issue or only part of the issue?

It doesn't really explain why forcing numpy to use a different malloc implementation would help.

**Comment by yaroslavvb** [comment]
That issue was about doing `a+=b` where `a` comes from Ray. It's expected to be slow because a is read-only so an extra copy is required.

This issue is about it being slow to load `a` onto GPU.  Both PyTorch and TensorFlow use `cudaMemCopyAsync` to copy things onto GPU, perhaps `cudaMemCopyAsync` has a problem with the kind of memory Ray provides

**Comment by yaroslavvb** [comment]
Actually, cudaMemCopyAsync is not the issue, it seems like both frameworks are doing single threaded memcpy before calling that. Not sure why copy is necessary, will open issue on PyTorch forums

**Comment by yaroslavvb** [comment]
OK, it seems both PyTorch and TensorFlow have a problem with read-only numpy arrays

Relevant issues:
https://github.com/tensorflow/tensorflow/issues/17233#issuecomment-369053165
https://github.com/pytorch/pytorch/issues/5442

This should be fixed on framework side, but meanwhile perhaps there's a way to have a workaround with Ray to allow writeable

**Comment by yaroslavvb** [comment]
It seems in some cases you can avoid extra copy by doing
```arr.flags['WRITEABLE'] = True```

This seems to regain the speed for PyTorch in all cases I tried. For TensorFlow, it works with TensorFlow 1.5 on single-socket/multi-socket system.

Latest TensorFlow seems to have a problem with feeding on multi-socket systems in general (https://github.com/tensorflow/tensorflow/issues/17339)

Also, for single-socket system, feed_dict is slow on ray arrays despite WRITEABLE=True. I wonder if it's a side-effect of being compiled with AVX

**Comment by robertnishihara** [comment]
Should be resolved now.

**Comment by ericl** [comment]
I'm seeing this again in TF 1.9 -- feeding writeable memory runs a 5GB/s, and read-only memory at <1GB/s.

**Comment by ericl** [comment]
Nevermind, it's fixed in later versions.
